### PR TITLE
C++: IR return indirections for `this`

### PR DESCRIPTION
**Pull request by rdmarsh2**

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll [hunk 1]
```diff
@@ -74,7 +74,12 @@ class ReturnValueNode extends ReturnNode {
 class ReturnIndirectionNode extends ReturnNode {
   override ReturnIndirectionInstruction primary;
 
-  override ReturnKind getKind() { result = TIndirectReturnKind(primary.getParameter().getIndex()) }
+  override ReturnKind getKind() {
+    result = TIndirectReturnKind(-1) and
+    primary.isThisIndirection()
+    or
+    result = TIndirectReturnKind(primary.getParameter().getIndex())
```
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:81]: I think we'd want to do both of those, and make the same `getASTParameter` change to the initialization instructions and argument operands. That's enough that I'd rather merge this as-is and do a sweeping parameter change separately.

**Comment by jbj** [comment]
I re-triggered the tests, and they are now failing. I'm guessing it's because of semantic merge conflicts with #3602 (trivial) and #3123 (complicated).

**Comment by rdmarsh2** [comment]
Yes, semantic merge conflicts. All the merge changes are positive, with the exception of a few conversions without locations being treated as sinks. I've removed those sinks in a separate commit.

**Comment by jbj** [comment]
I've started https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1181/, hoping that our Jenkins workers are in better shape this morning than they were last night.

The test changes look great! I hope the CPP-Differences will live up to that.

**Comment by rdmarsh2** [comment]
The new `cpp/unbounded-write` results in the CPP-Differences are all false positives - the length of the `strcpy` is bounded because the output argument of `fgets` is bounded. The added edges all look accurate, though.

**Comment by jbj** [comment]
LGTM, but there's a merge conflict.

**Pull request merged**